azkaban-exec-server/src/main/java/azkaban/execapp/JobRunner.java

```diff
@@ -739,29 +741,33 @@ private Status runJob() {
     try {
       this.job.run();
     } catch (final Throwable e) {
-      if (this.props.getBoolean("job.succeed.on.failure", false)) {
-        finalStatus = changeStatus(Status.FAILED_SUCCEEDED);
-        logError("Job run failed, but will treat it like success.");
-        logError(e.getMessage() + " cause: " + e.getCause(), e);
-      } else {
-        if (isKilled() || this.node.getStatus() == Status.KILLED) {
-          finalStatus = Status.KILLED;
-          logError("Job run killed!", e);
+      synchronized (this.syncObject) {
+        if (this.props.getBoolean("job.succeed.on.failure", false)) {
+          finalStatus = changeStatus(Status.FAILED_SUCCEEDED);
+          logError("Job run failed, but will treat it like success.");
+          logError(e.getMessage() + " cause: " + e.getCause(), e);
         } else {
-          finalStatus = changeStatus(Status.FAILED);
-          logError("Job run failed!", e);
+          if (isKilled() || this.node.getStatus() == Status.KILLED) {
+            finalStatus = Status.KILLED;
+            logError("Job run killed!", e);
+          } else {
+            finalStatus = changeStatus(Status.FAILED);
+            logError("Job run failed!", e);
+          }
+          logError(e.getMessage() + " cause: " + e.getCause());
         }
-        logError(e.getMessage() + " cause: " + e.getCause());
       }
     }
 
     if (this.job != null) {
       this.node.setOutputProps(this.job.getJobGeneratedProperties());
     }
 
-    // If the job is still running, set the status to Success.
-    if (!Status.isStatusFinished(finalStatus)) {
-      finalStatus = changeStatus(Status.SUCCEEDED);
+    synchronized (this.syncObject) {
+      // If the job is still running, set the status to Success.
+      if (!Status.isStatusFinished(finalStatus) && !this.killed) {
```

Comment:
> And if killed is true, who will update the final status?

Here (KILLED is a "final status", it doesn't have to be set again after this):
https://github.com/juhoautio/azkaban/blob/723a46ec93d6e52f178e5d2ab445d630099d0ad6/azkaban-exec-server/src/main/java/azkaban/execapp/JobRunner.java#L833

Although now that you mention it, I'm not sure what would happen if this row is hit:
https://github.com/juhoautio/azkaban/blob/723a46ec93d6e52f178e5d2ab445d630099d0ad6/azkaban-exec-server/src/main/java/azkaban/execapp/JobRunner.java#L822 – seems a bit weird, I think we should add `this.changeStatus(Status.KILLED)` before exiting with this special `return`.